### PR TITLE
End the reign of RoutingActivity

### PR DESCRIPTION
**Pull request by mcginty**
Apparently you can't re-open a PR after you've force-pushed so had to make a new one. Reference https://github.com/WhisperSystems/TextSecure/pull/2331 for previous comments.

### `Intent` logic
* Move routing logic to `PassphraseRequiredActionBarActivity` which routes based on application state just as `RoutingActivity` once did in days of olde.
* Children activities now implement `onCreate(Bundle bundle, MasterSecret masterSecret)`, which is only called if we don't have to route elsewhere.
* ShareActivity now handles `SEND` intents
* ConversationListActivity is now the launcher entry-point
* ConversationActivity handles SMS `SENDTO` intents

### `MasterSecret`
* `MasterSecret`s are initialized in `PassphraseRequiredActionBarActivity`, and accessed via protected getter instead of being passed inter-Activity via parceled secrets. This means one MasterSecret object per activity that is cleared `onDestroy()`, so less chance of leak since fewer copies are made.
* `MemoryCleaner`'s code was commented out, so just removed for now.

Can't reproduce #2239 on this.

444-line net :small_red_triangle_down:!

### CODE REVIEW

File: src/org/thoughtcrime/securesms/ApplicationPreferencesActivity.java [hunk 1]
```diff
@@ -129,18 +126,19 @@ public void onCreate(Bundle icicle) {
       super.onCreate(icicle);
       addPreferencesFromResource(R.xml.preferences);
 
+      MasterSecret masterSecret = getArguments().getParcelable("master_secret");
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/ApplicationPreferencesActivity.java]: what is this?

File: src/org/thoughtcrime/securesms/ConversationActivity.java [hunk 2]
```diff
@@ -197,6 +200,32 @@ protected void onNewIntent(Intent intent) {
...
+
+      String data = intent.getData().getSchemeSpecificPart();
+      recipients = RecipientFactory.getRecipientsFromString(this, data, false);
+      threadId   = DatabaseFactory.getThreadDatabase(this).getThreadIdIfExistsFor(recipients);
+
+      if (recipients == null || recipients.isEmpty()) {
+        Toast.makeText(this, R.string.ConversationActivity_specify_recipient, Toast.LENGTH_LONG).show();
+        Intent selectIntent = new Intent(this, NewConversationActivity.class);
+        selectIntent.putExtra(DRAFT_TEXT_EXTRA, body);
+        startActivity(selectIntent);
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/ConversationActivity.java]: this is still weird

File: src/org/thoughtcrime/securesms/ConversationActivity.java [hunk 3]
```diff
@@ -162,15 +161,17 @@
...
-  protected void onCreate(Bundle state) {
+  protected void onCreate(Bundle state, MasterSecret masterSecret) {
+    this.masterSecret = masterSecret;
     overridePendingTransition(R.anim.slide_from_right, R.anim.fade_scale_out);
     dynamicTheme.onCreate(this);
     dynamicLanguage.onCreate(this);
-    super.onCreate(state);
 
     setContentView(R.layout.conversation_activity);
+    ((ConversationFragment) getSupportFragmentManager().findFragmentById(R.id.fragment_content)).setMasterSecret(masterSecret);
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/ConversationActivity.java]: why not put this in the fragment's bundle?

File: src/org/thoughtcrime/securesms/DatabaseUpgradeActivity.java [hunk 4]
```diff
@@ -38,6 +38,7 @@
 import org.thoughtcrime.securesms.jobs.PushDecryptJob;
 import org.thoughtcrime.securesms.jobs.SmsDecryptJob;
 import org.thoughtcrime.securesms.notifications.MessageNotifier;
+import org.thoughtcrime.securesms.service.KeyCachingService;
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/DatabaseUpgradeActivity.java]: ?

File: src/org/thoughtcrime/securesms/GroupCreateActivity.java [hunk 5]
```diff
@@ -95,7 +95,8 @@
 
   public static final String GROUP_RECIPIENT_EXTRA = "group_recipient";
   public static final String GROUP_THREAD_EXTRA    = "group_thread";
-  public static final String MASTER_SECRET_EXTRA   = "master_secret";
+
+  private MasterSecret masterSecret;
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/GroupCreateActivity.java]: ?

File: src/org/thoughtcrime/securesms/MediaOverviewActivity.java [hunk 6]
```diff
@@ -63,11 +63,9 @@
   private long              threadId;
 
   @Override
-  protected void onCreate(Bundle bundle) {
+  protected void onCreate(Bundle bundle, MasterSecret masterSecret) {
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/MediaOverviewActivity.java]: you have an instance variable but don't seem to set it?

Comment (moxie0) [inline src/org/thoughtcrime/securesms/MediaOverviewActivity.java]: also do you still need MASTER_SECRET_EXTRA?

File: src/org/thoughtcrime/securesms/PassphraseRequiredActionBarActivity.java [hunk 7]
```diff
@@ -1,29 +1,52 @@
...
+  private static final int STATE_PROMPT_PASSPHRASE        = 2;
+  private static final int STATE_UPGRADE_DATABASE         = 3;
+  private static final int STATE_PROMPT_PUSH_REGISTRATION = 4;
 
   private final PassphraseRequiredMixin delegate = new PassphraseRequiredMixin();
 
+  private boolean isVisible;
+
   @Override
   protected void onCreate(Bundle savedInstanceState) {
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/PassphraseRequiredActionBarActivity.java]: final

File: src/org/thoughtcrime/securesms/notifications/MarkReadReceiver.java [hunk 8]
```diff
@@ -9,6 +9,7 @@
 
 import org.thoughtcrime.securesms.crypto.MasterSecret;
 import org.thoughtcrime.securesms.database.DatabaseFactory;
+import org.thoughtcrime.securesms.service.KeyCachingService;
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/notifications/MarkReadReceiver.java]: ?

File: src/org/thoughtcrime/securesms/ApplicationPreferencesActivity.java (hunk 1 above)
Comment (mcginty) [inline src/org/thoughtcrime/securesms/ApplicationPreferencesActivity.java]: The MasterSecret is passed in as part of the arguments bundle, so we unparcel it here and pass it through to fragments that need it (in this case the app protection screen fragment).

File: src/org/thoughtcrime/securesms/MediaOverviewActivity.java (hunk 6 above)
Comment (mcginty) [inline src/org/thoughtcrime/securesms/MediaOverviewActivity.java]: missed on rebase, fixed.

File: src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java [hunk 9]
```diff
@@ -126,7 +126,7 @@ public boolean onPreferenceChange(final Preference preference, Object newValue)
         builder.setPositiveButton(R.string.ApplicationPreferencesActivity_disable, new DialogInterface.OnClickListener() {
           @Override
           public void onClick(DialogInterface dialog, int which) {
-            MasterSecret masterSecret = getActivity().getIntent().getParcelableExtra("master_secret");
+            MasterSecret masterSecret = KeyCachingService.getMasterSecret(getActivity());
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/preferences/AppProtectionPreferenceFragment.java]: the only thing that should reference KeyCachingService is the passphrase activity base class

**Comment by McLoo** [comment]
@mcginty isn't it possible to move all/most dynamicLanguage and dynamicTheme to PassphraseRequiredActionBarActivit.OnCreate() or even BaseActionBarActivity?

**Comment by mcginty** [comment]
@McLoo I'd like to do that in another change. Not all activities use DynamicTheme and I don't want to break themes or feature-creep this PR.

**Comment by mcginty** [comment]
@McLoo Also that NPE you posted earlier looked unrelated to this change, so going to have a separate PR for that as well.

**Comment by mcginty** [comment]
(known bug with opening messages via notification - will fix tomorrow)

File: src/org/thoughtcrime/securesms/ConversationListFragment.java [hunk 10]
```diff
@@ -114,7 +111,6 @@ public void onResume() {
   @Override
   public void onAttach(Activity activity) {
     super.onAttach(activity);
-    this.listener = (ConversationSelectedListener)activity;
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/ConversationListFragment.java]: still need the method?

File: src/org/thoughtcrime/securesms/ShareActivity.java [hunk 11]
```diff
@@ -102,10 +99,14 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
-    startActivity(new Intent(this, RoutingActivity.class));
-    super.onMasterSecretCleared();
+  private void initializeFragment() {
+    Bundle args = new Bundle();
+    args.putParcelable("master_secret", masterSecret);
+    final ShareFragment fragment = new ShareFragment();
+    fragment.setArguments(args);
+    getSupportFragmentManager().beginTransaction()
+                               .replace(android.R.id.content, fragment)
+                               .commit();
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/ShareActivity.java]: Can't this pattern be a helper method on the base class?  initializeFragment(android.R.id.content, new ShareFragment());

File: src/org/thoughtcrime/securesms/PassphraseRequiredActionBarActivity.java [hunk 12]
```diff
@@ -1,42 +1,186 @@
...
   public void onMasterSecretCleared() {
-    finish();
+    Log.w(TAG, "onMasterSecretCleared()");
+    if (isVisible) routeApplicationState(null);
+    else           finish();
   }
 
-  @Override
-  public void onNewMasterSecret(MasterSecret masterSecret) {}
+  protected <T extends Fragment> T initFragment(@IdRes int target,
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/PassphraseRequiredActionBarActivity.java]: protected?

File: src/org/thoughtcrime/securesms/PassphraseRequiredActionBarActivity.java [hunk 13]
```diff
@@ -1,42 +1,186 @@
...
+    Bundle args = new Bundle();
+    args.putParcelable("master_secret", masterSecret);
+    fragment.setArguments(args);
+    getSupportFragmentManager().beginTransaction()
+                               .replace(target, fragment)
+                               .commit();
+    return fragment;
+  }
+
+  protected void routeApplicationState(MasterSecret masterSecret) {
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/PassphraseRequiredActionBarActivity.java]: protected?

File: src/org/thoughtcrime/securesms/PassphraseRequiredActionBarActivity.java [hunk 14]
```diff
@@ -1,42 +1,186 @@
...
+    Intent intent = getIntentForState(masterSecret, getApplicationState(masterSecret));
+    if (intent != null) {
+      startActivity(intent);
+      finish();
+    }
+  }
+
+  protected Intent getIntentForState(MasterSecret masterSecret, int state) {
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/PassphraseRequiredActionBarActivity.java]: protected?

Comment (mcginty) [inline src/org/thoughtcrime/securesms/PassphraseRequiredActionBarActivity.java]: Was going to offer them to children activities as I thought they'd be useful, but they're not that useful. Making private.

**Comment by moxie0** [comment]
ok lgtm, i'll put it in 2.10.0

**Pull request closed without merging**